internal/generate/servicepackage/file.gtpl

```diff
@@ -132,11 +135,40 @@ func (p *servicePackage) ServicePackageName() string {
 // NewClient returns a new AWS SDK for Go v2 client for this service package's AWS API.
 func (p *servicePackage) NewClient(ctx context.Context, config map[string]any) (*{{ .GoV2Package }}.Client, error) {
 	cfg := *(config["aws_sdkv2_config"].(*aws.Config))
-
-	return {{ .GoV2Package }}.NewFromConfig(cfg,
+	optFns := []func(*{{ .GoV2Package }}.Options){
 		{{ .GoV2Package }}.WithEndpointResolverV2(newEndpointResolverV2()),
 		withBaseEndpoint(config[names.AttrEndpoint].(string)),
-	), nil
+{{- if gt (len .EndpointRegionOverrides) 0 }}
+		func(o *{{ .GoV2Package }}.Options) {
+			switch partition := config["partition"].(string); partition {
+	{{- range $k, $v := .EndpointRegionOverrides }}
+			case endpoints.{{ $k | Camel }}PartitionID:
+				if region := endpoints.{{ $v | Camel }}RegionID; cfg.Region != region {
+					tflog.Info(ctx, "overriding region", map[string]any{
+						"original_region": cfg.Region,
+						"override_region": region,
+					})
+					o.Region = region
+				}
+	{{- end }}
+			}
+		},
+{{- end }}
+	}
+
+	optFns = append(optFns, servicePackageExtraOptFns(ctx, p, config)...)
+
+	return {{ .GoV2Package }}.NewFromConfig(cfg, optFns...), nil
+}
+
+func servicePackageExtraOptFns(ctx context.Context, sp conns.ServicePackage, config map[string]any) []func(*{{ .GoV2Package }}.Options) {
+	if v, ok := sp.(interface {
+		extraOptFns(context.Context, map[string]any) []func(*{{ .GoV2Package }}.Options)
+	}); ok {
+		return v.extraOptFns(ctx, config)
+	}
+
+	return nil
```

Comment:
Will these "extra" options be exported and referenced elsewhere in the future? I find the client setup a bit harder to read with this function called via a separate `append` operation, but understand if the tradeoff is necessary to future proof for multi-region support.

Reply:
The immediate reason is so that we can generate most of the API client factory method for those dozen or so API services that are doing it manually -- they are doing that because of setting up custom retry handling which is done via an `Options` function. This will allow that functionality to be moved into a per-service `extraOptFns` method.
Agreed that the `append` is not nice, would help if you could do something like this in Go:

```go
optFns := []func(*Options) {
    blah,
   extraOptFns(ctx, config)...,
}
```

Reply:
Realistically we could just have at most one extra `OptFn` per service and insert it in the slice initializer and instead of returning `nil` by default, return a no-op `OptFn`.

Reply:
Added `withExtraOptions` which can now be called directly in the slice initializer.